File: python/OrphanRef.py
from __future__ import annotations

# coding: utf-8
# OrphanRef.py
# 孤立したサンプルをimputeする

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
import Orphan
import RefCommon
from KnownFamily import *
from OptionSmall import OptionSmall


#################### OrphanRef ####################

def impute(samples: list[str], orig_vcf: VCFSmall, ref_haps: list[list[int]],
				phased_vcf: VCFGeno, op: OptionSmall) -> Optional[VCFGenoBase]:
	if not samples:
		return None
	
	vcf = VCFGeno.extract_samples(samples, orig_vcf);
	records = RefCommon.expand_records(vcf, phased_vcf)
	vcf1 = Orphan.impute_samples(samples, records, ref_haps, orig_vcf, op)
	if vcf1 is not None:
		print("%d orphan samples have been imputed." % len(samples))
	return vcf1

__all__ = ['impute']

File: src/OrphanRef.cpp
#include "../include/OrphanRef.h"
#include "../include/Orphan.h"
#include "../include/VCFGeno.h"
#include "../include/RefCommon.h"

using namespace std;


//////////////////// OrphanRef ////////////////////

VCFGenoBase *OrphanRef::impute(const std::vector<std::string>& samples,
								const VCFSmall *orig_vcf,
								const std::vector<std::vector<int>>& ref_haps,
								const VCFGeno *phased_vcf,
								const OptionSmall& op) {
	if(samples.empty())
		return NULL;
	
	auto	*vcf = VCFGeno::extract_samples(samples, orig_vcf);
	const auto	records = RefCommon::expand_records(vcf, phased_vcf);
	auto	*vcf1 = Orphan::impute_samples(samples, records,
											ref_haps, orig_vcf, op);
	if(vcf1 != NULL)
		cerr << samples.size() << " orphan samples have been imputed." << endl;
	delete vcf;
	return vcf1;
}
